### PR TITLE
Feature/initial pr

### PR DESCRIPTION
**Pull request by dbernstein**
Initial pull request for the palace-quicksight command-line tool.

It includes 3 functions: 

1. Export a Quicksight Analysis and related data sets  as a Quicksight template in the json format.
2. Import a previously exported Quicksight template from json and hook it up to a data source of choice
3. Publish a previously imported Template to a dashboard.

The process has been tested manually in the tpp-dev quicksight account.

You can run those tests against the dev environment yourself by running the following commands:

## Export Analysis
```
poetry run ./bin/palace-quicksight export-analysis --aws-profile palace-dev --aws-account-id 128682227026 --analysis-id d954330e-7b80-4a4a-ab64-47f53d8eea38 --output-dir /path/to/output/dir
```
## Import Template
```
poetry run ./bin/palace-quicksight import-template --aws-profile palace-dev --aws-account-id 128682227026 --template-name "library" --data-source-arn arn:aws:quicksight:us-west-2:128682227026:datasource/2d70d5ea-1143-4de4-adcb-0dbce7ce0d6e --input-dir /path/to/output/dir --target-namespace tpp-test-env
```
## Publish Template
```
poetry run ./bin/palace-quicksight  publish-dashboard --aws-profile palace-dev --aws-account-id 128682227026 --template-id "tpp-test-env-library" --group-name readers --target-namespace tpp-test-env
```

### CODE REVIEW

File: core/operation/export_analysis_operation.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            self._template_definition = self._get_template_definition(
+                template_id=template_id
+            )
+
+            return "SUCCESSFUL" in self._template_definition["ResourceStatus"]
+
+        retry(verify_success)
+
+        # get the newly created template definition
+        self._log.info(f"Writing template definition response to disk")
```
Comment (RishiDiwanTT) [inline core/operation/export_analysis_operation.py:71]: Can we add the filename to the log, so it is easy to find for the user.

File: core/operation/import_from_json_operation.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        target_namespace: str,
+        data_source_arn: str,
+        input_dir: str,
+        *args,
+        **kwargs,
+    ):
+        self._template_name = template_name
+        self._target_namespace = target_namespace
+        self._data_source_arn = data_source_arn
+        self._intput_dir = input_dir
```
Comment (RishiDiwanTT) [inline core/operation/import_from_json_operation.py:24]: typo

File: core/operation/import_from_json_operation.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    def execute(self):
+        # Read template file into dictionary
+        template_data = None
+        template_file = self._resolve_path(
+            self._intput_dir, TEMPLATE_DIR, self._template_name + ".json"
+        )
+        with open(template_file) as template_file:
+            template_data = json.loads(template_file.read())
+
+        # create namespace if not exists
```
Comment (RishiDiwanTT) [inline core/operation/import_from_json_operation.py:36]: I assume this is to be deleted?

File: core/operation/import_from_json_operation.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    def _create_or_update_data_set(self, dataset_definition: dict):
+        """
+        Create new or updates existing DataSet
+        :param dataset_definition:
+        :return: DataSet ARN and DataSet Id
+        """
+        try:
+            response = self._qs_client.create_data_set(**dataset_definition)
+        except self._qs_client.exceptions.ResourceExistsException as e:
+            response = self._qs_client.update_data_set(**dataset_definition)
```
Comment (RishiDiwanTT) [inline core/operation/import_from_json_operation.py:94]: I did not check, but maybe the create and update APIs have differing parameters here as well?

File: core/util.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+from time import sleep
+
+
+def retry(func) -> bool:
```
Comment (RishiDiwanTT) [inline core/util.py:4]: There are libraries that provide configurable methods of retrying till success, eg. https://github.com/invl/retry
If we don't need the configurability of these libraries we can stick to this implementation.

File: core/util.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+def recursively_replace_value(mydict: dict, key: str, val: str):
+    """
+    Recursively searches mydict for key and replaces its value (if any) with val
+    :param mydict:
+    :param key:
+    :param val:
+    :return:
+    """
+    if key in mydict:
```
Comment (RishiDiwanTT) [inline core/util.py:28]: Imaginary edge case: What if `key` is not a `str` but a list or dict, should we safeguard against replacing such values?

File: tests/core/operation/test_export_analysis_operation.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    describe_template_definition_response,
+    get_analysis_definition_response,
+    get_analysis_description_response,
+)
+
+
+class TestExportAnalysisOperation:
+    def test(self):
+        analysis_id = "my-quicksight-analysis-id"
+        output_dir = "/tmp/test-output"
```
Comment (RishiDiwanTT) [inline tests/core/operation/test_export_analysis_operation.py:21]: A more platform independent API exists under https://docs.python.org/3/library/tempfile.html
Or you can use the mocking strategy https://docs.python.org/3/library/unittest.mock.html#mock-open

**Review by RishiDiwanTT — COMMENTED** [review]
I've added some feedback which is mostly about code style or alternative APIs. Logically everything seems fine :+1:

File: core/cli.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    return boto3.client("quicksight")
+
+
+@click.group()
+def cli():
+    pass
+
+
+@click.command()
+@click.option("--aws-profile", required=True, help="The AWS account profile")
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply core/cli.py:26]: I suppose because I wanted to require that the user to make an explicit choice - ie I don't want default aws configs to be used by mistake.

File: core/cli.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    required=True,
+    help="The path to the output directory to which resources will be exported",
+)
+def export_analysis(
+    aws_profile: str, aws_account_id: str, analysis_id: str, output_dir: str
+):
+    """
+    Exports a template and dependent data sets based on the specified analysis to JSON files.
+    """
+    click.echo(f"Create version")
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply core/cli.py:42]: Yes. Good idea.

File: core/operation/baseoperation.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    def _create_or_update_template(self, template_data: dict) -> tuple[str, str, str]:
+        """
+        Creates new or updates existing template.
+        :param template_data:
+        :return: Template ARN, Template Version ARN, and the Template ID
+        """
+        try:
+            response = self._qs_client.create_template(**template_data)
+        except self._qs_client.exceptions.ResourceExistsException as e:
+            response = self._qs_client.update_template(**template_data)
```
[the remark this answers is not in the document]
Comment (dbernstein) [reply core/operation/baseoperation.py:35]: I think it makes more sense to delete the template (or dataset) if it exists and recreate it. That way we don't have to deal with tags and permissions updates separately.

File: core/operation/import_from_json_operation.py (hunk 4 above)
Comment (dbernstein) [reply core/operation/import_from_json_operation.py:98]: Same as above: I will fix by deleting and re-creating.

File: core/util.py (hunk 6 above)
Comment (dbernstein) [reply core/util.py:28]: It's case I don't think we need to worry about.

File: core/util.py (hunk 5 above)
Comment (dbernstein) [reply core/util.py:4]: For the time being I don't think we need configurability.

File: core/cli.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dbernstein) [reply core/cli.py:33]: But in looking more closely at this,  there really is no good reason not to use AWS_PROFILE or other standard approaches.  Plus it fits better with our terraform usage.  I will fix.

**Review by dbernstein — COMMENTED** [review]
@RishiDiwanTT :  Thanks for the great feedback.  I've updated the PR with your suggested changes.  I also added tests to cover import and publish.

**Comment by dbernstein** [comment]
@RishiDiwanTT :  can you give the updates a quick look:  if all is well, approve it and I'll merge it.

File: core/operation/import_from_json_operation.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+            self._qs_client.delete_data_set(
+                **{
+                    "DataSetId": dataset_definition["DataSetId"],
+                    "AwsAccountId": dataset_definition["AwsAccountId"],
+                }
+            )
+
+            # there can be some latency between the completion of the deletion command
+            # and the complete backend deletion operation.
+            time.sleep(3)
```
Comment (RishiDiwanTT) [inline core/operation/import_from_json_operation.py:112]: Maybe we should add a log here so we can track when resources were deleted

**Review by RishiDiwanTT — APPROVED** [review]
One more tiny comment, all good :+1:
